components/View/Taebul/index.js

```diff
@@ -26,32 +31,172 @@ const defaultProps = {
     settings: {},
 };
 
+const MAX_IDLE_TIMEOUT = 200;
+
+const getVirtualizedRenderParams = (
+    columns,
+    scrollOffset,
+    width,
+    itemWidths,
+) => {
+    let i = 0;
+    let startX = 0;
+    let startIndex;
+
+    for (; i < columns.length; i += 1) {
+        const newStartX = startX + itemWidths[columns[i].key];
+        startIndex = i;
+
+        if (scrollOffset < newStartX) {
+            break;
+        }
+
+        startX = newStartX;
+    }
+
+    let endIndex = startIndex;
+    let endX = startX;
+
+    for (; i < columns.length; i += 1) {
+        const newEndX = endX + itemWidths[columns[i].key];
+        endIndex = i;
+        endX = newEndX;
+
+        // NOTE: 10px to accomodate the scrollbar
+        if ((width + (scrollOffset - 10)) < newEndX) {
+            break;
+        }
+    }
+
+    return {
+        startIndex,
+        endIndex,
+        startX,
+        endX,
+    };
+};
+
+const emptyObject = {};
+
 export default class Taebul extends React.PureComponent {
     static propTypes = propTypes;
     static defaultProps = defaultProps;
 
     static columnKeySelector = column => column.key;
 
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            scrollLeft: 0,
+        };
+
+        // this.calculateRowVirtualizationParams(props);
+    }
+
     componentDidMount() {
         window.addEventListener('scroll', this.handleScroll, true);
     }
 
+    componentWillReceiveProps(nextProps) {
+        const { scrollLeft } = this.state;
+        this.calculateRowVirtualizationParams(nextProps, scrollLeft);
+    }
+
     componentWillUnmount() {
         window.removeEventListener('scroll', this.handleScroll, true);
     }
 
+    getTotalWidth = memoize((columns, columnWidths, defaultColumnWidth) => {
+        const definedColumnWidthList = Object.values(columnWidths).filter(isDefined);
+        const definedColumnWidthSum = sum(definedColumnWidthList);
+
+        const defaultColumnWidthSum = (columns.length - definedColumnWidthList.length)
+            * defaultColumnWidth;
+
+        return defaultColumnWidthSum + definedColumnWidthSum;
+    })
+
+
+    getItemWidths = memoize((
+        columns,
+        columnWidths,
+        defaultColumnWidth,
+    ) => {
+        const itemWidths = {};
+        columns.forEach((c) => {
+            itemWidths[c.key] = columnWidths[c.key] || defaultColumnWidth;
```

Comment:
This is potentially dangerous, columnWidths can be 0, in which case it will be expanded to defaultColumnWidth